testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketConnectTest.java

```diff
@@ -110,8 +124,80 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        run();
+    }
+
+    public void testWriteWithFastOpenBeforeConnect(ServerBootstrap sb, Bootstrap cb) throws Throwable {
+        enableTcpFastOpen(sb, cb);
+
+        ServerBootstrapHandler handler = new ServerBootstrapHandler();
+        sb.childHandler(handler);
+
+        Channel sc = sb.bind().sync().channel();
+        connectAndVerifyDataTransfer(cb, sc, handler);
+        connectAndVerifyDataTransfer(cb, sc, handler);
+    }
+
+    private static void connectAndVerifyDataTransfer(Bootstrap cb, Channel sc, ServerBootstrapHandler handler)
+            throws InterruptedException {
+        cb.handler(new ChannelInboundHandlerAdapter());
+        ChannelFuture register = cb.register();
+        Channel channel = register.sync().channel();
+        ChannelFuture write = channel.write(writeAscii(DEFAULT, "[fastopen]"));
+        SocketAddress remoteAddress = sc.localAddress();
+        ChannelFuture connectFuture = channel.connect(remoteAddress);
+        Channel cc = connectFuture.sync().channel();
+        cc.writeAndFlush(writeAscii(DEFAULT, "[normal data]")).sync();
+        write.sync().isDone();
+        cc.disconnect().sync();
+        String expectedString = "[fastopen][normal data]";
+        String result = handler.collectBuffer(expectedString.getBytes(US_ASCII).length);
+        assertEquals(expectedString, result);
+    }
+
+    protected void enableTcpFastOpen(ServerBootstrap sb, Bootstrap cb) {
+        throw new AssumptionViolatedException(
+                "Support for testing TCP_FASTOPEN not enabled for " + StringUtil.simpleClassName(this));
+    }
+
     private static void assertLocalAddress(InetSocketAddress address) {
         assertTrue(address.getPort() > 0);
         assertFalse(address.getAddress().isAnyLocalAddress());
     }
+
+    @ChannelHandler.Sharable
+    private static class ServerBootstrapHandler extends ChannelInboundHandlerAdapter {
+        private final Semaphore semaphore = new Semaphore(0);
+        private final StringBuffer stringBuffer = new StringBuffer();
+
+        @Override
+        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+            ctx.read();
+            super.channelActive(ctx);
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            if (msg instanceof ByteBuf) {
+                ByteBuf buf = (ByteBuf) msg;
+                int readableBytes = buf.readableBytes();
+                stringBuffer.append(buf.readCharSequence(readableBytes, US_ASCII));
+                semaphore.release(readableBytes);
+            } else {
+                throw new IllegalArgumentException("Unexpected message type: " + msg);
+            }
+            super.channelRead(ctx, msg);
+        }
+
+        String collectBuffer(int expectedBytes) throws InterruptedException {
+            semaphore.acquire(expectedBytes);
+            synchronized (stringBuffer) {
```

Comment:
why the synchronized ?

Reply:
right, no need for anymore now that the semaphore counts bytes